src/server/events/execution_event_bus.ts

```diff
@@ -13,16 +11,165 @@ export interface ExecutionEventBus {
+  // Structure: eventName -> listener -> array of wrapped listeners (to handle multiple registrations)
+  private listenerMap: Map<'event' | 'finished', Map<Listener, WrappedListener[]>> = new Map();
+
   constructor() {
     super();
+    this.listenerMap.set('event', new Map());
+    this.listenerMap.set('finished', new Map());
   }
 
   publish(event: AgentExecutionEvent): void {
-    this.emit('event', event);
+    this.dispatchEvent(new CustomEventImpl('event', { detail: event }));
   }
 
   finished(): void {
-    this.emit('finished');
+    this.dispatchEvent(new Event('finished'));
+  }
+
+  /**
+   * EventEmitter-compatible 'on' method.
+   * Wraps the listener to extract event detail from CustomEvent.
+   * Supports multiple registrations of the same listener (like EventEmitter).
+   */
+  on(eventName: 'event' | 'finished', listener: Listener): this {
+    const wrappedListener: WrappedListener = (e: Event) => {
+      if (e.type === 'event') {
+        listener((e as CustomEvent<AgentExecutionEvent>).detail);
+      } else {
+        // 'finished' event - call with undefined (matches EventEmitter behavior)
+        listener(undefined as unknown as AgentExecutionEvent);
+      }
+    };
+
+    const eventListeners = this.listenerMap.get(eventName)!;
+    const wrappedListeners = eventListeners.get(listener);
+    if (wrappedListeners) {
+      wrappedListeners.push(wrappedListener);
+    } else {
+      eventListeners.set(listener, [wrappedListener]);
+    }
+    this.addEventListener(eventName, wrappedListener);
+    return this;
+  }
+
+  /**
+   * EventEmitter-compatible 'off' method.
+   * Uses the stored wrapped listener for proper removal.
+   * Removes one instance at a time (LIFO order, like EventEmitter).
+   */
+  off(eventName: 'event' | 'finished', listener: Listener): this {
+    const eventListeners = this.listenerMap.get(eventName)!;
+    const wrappedListeners = eventListeners.get(listener);
+    if (wrappedListeners && wrappedListeners.length > 0) {
+      // Remove the most recently added listener (LIFO)
+      const wrappedListener = wrappedListeners.pop()!;
+      this.removeEventListener(eventName, wrappedListener);
+      // Clean up the map entry if no more wrapped listeners
+      if (wrappedListeners.length === 0) {
+        eventListeners.delete(listener);
+      }
+    }
+    return this;
+  }
+
+  /**
+   * EventEmitter-compatible 'once' method.
+   * Listener is automatically removed after first invocation.
+   * Supports multiple registrations of the same listener (like EventEmitter).
+   */
+  once(eventName: 'event' | 'finished', listener: Listener): this {
+    const eventListeners = this.listenerMap.get(eventName)!;
+
+    const wrappedListener: WrappedListener = (e: Event) => {
+      // Clean up tracking for this specific wrapped listener
+      const wrappedListeners = eventListeners.get(listener);
+      if (wrappedListeners) {
```

Comment:
nit: (to be the same as in `off`)
```suggestion
      if (wrappedListeners && wrappedListeners.length > 0) {
```